util/string_ref.c

```diff
@@ -0,0 +1,125 @@
+/*
+   Copyright 2020 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include <plhash.h>
+#include <assert.h>
+#include <pthread.h>
+#include "string_ref.h"
+#include "locks_wrap.h"
+#include "comdb2_atomic.h"
+#include "logmsg.h"
+
+#define TRACK_REFERENCES
+
+#ifdef TRACK_REFERENCES
+static hash_t *hrefs = NULL;
+static pthread_mutex_t hmtx = PTHREAD_MUTEX_INITIALIZER;
+#endif
+int gbl_creation_count;
+
+struct string_ref_t {
```

Comment:
A common approach for this would be:

```c
struct string_ref_t {
    int cnt;
    char str[1];
};


string_ref_t *ref = malloc(sizeof(string_ref_t) + strlen(src));
strcpy(ref->str, src);
```

Reply:
Yes, I will change that one.